### ISSUE TITLE
Weird results

### ISSUE THREAD

**Issue by ellick53** [rpautrat/SuperPoint#65]
Hello Rémi, and thanks for your awesome job.

For some reason, when I train MagicPoint using Synthetic Shapes, everything is fine. But when I train on COCO, there are points around the corners, see the image:

![Problem](https://user-images.githubusercontent.com/49749099/56349909-db796e80-61c9-11e9-8a8a-a6b66283a97d.png)

Same thing happens when training SuperPoint on COCO. I haven't changed the model or the core parameters (as NMS, data augmentation, etc.), so I can't figure out what is going on. 

Do you have any idea what is going on and how to solve this problem?

Thanks a lot!

**Comment by rpautrat** [comment]
Hi! I checked with the version of SuperPoint I trained and I don't exactly observe the same behavior (for me I have only one corner detected each time and sometimes two points for one corner).

There might be several reasons and tuning that can be made to correct this. First, it might be because the scale of the corners in COCO is different from the scale of the synthetic shapes. You can try to increase the parameter `scaling_amplitude` in `configs/magic-point_coco_train.yaml` to have a bigger scale diversity in your COCO images.

Second, the additional points probably mean that the distribution of probability of a corner is not as peaky as with MagicPoint trained on the synthetic dataset, but more like a blob with a higher variance. You can increase the parameter `detection_threshold` when predicting the corners with SuperPoint until only the right corners are detected (the additional points should have a lower probability than the correct points).

Third, you can also increase the NMS to remove points that are too close (but I would try this as a last resort, try the other two before).

**Comment by ellick53** [comment]
Thanks, setting the detection to about 0.1 (instead of the default 0.015) seems to have solved the problem, although I occasionally observe the "double points" problem.

As a side question, SuperPoint detects points better than MagicPoint, even if trained on the same dataset (COCO). Do you have any idea why?

**Comment by rpautrat** [comment]
In SuperPoint, the VGG encoder is shared between the detector and descriptor heads. So the backpropagation from the descriptors loss also modifies the encoder so that the same point in different images should have similar descriptors. Thus, I guess that it helps improving the repeatability of the detector, even though I am not quite sure how.

**Comment by ellick53** [comment]
The fact that I had to increase the threshold still bothers me, as it really looks like a workaround "hiding" another issue. I will try again from the beginning, and see how it goes.

By the way, I am using TensorFlow GPU 1.13.1 instead of 1.6. Could this be the cause of this behavior?

**Comment by rpautrat** [comment]
Yes, I think that increase the level of scaling amplitude is the "cleanest" way to improve your detections.

1. I don't think that it would change anything.
2. You could do for example (and it is what I did): MagicPoint Shapes -> Export -> Train MagicPoint on COCO -> New export -> Train SuperPoint. The "rounds of homographic adaptation" would be multiple steps "Train MagicPoint on COCO -> Export".
3. You can already download the weights and graph of my SuperPoint training here: https://github.com/mmmfarrell/SuperPoint/blob/master/pretrained_models/sp_v5.tgz. If you really need the original checkpoints of MagicPoint and SuperPoint, I can also send them by email.

**Comment by ellick53** [comment]
Thanks a lot, I'm going to check it out!

**Comment by rpautrat** [comment]
There is an example of how to use the .pb file in https://github.com/mmmfarrell/SuperPoint/blob/master/superpoint/match_features_demo.py (this branch will be merged with the master branch in the future). But I will also send you the checkpoints.

The exports are actually the homographic adaptations: each time you export with the parameter homographic_adaptation->num: 100, it means that you use the homographic adaptation. The first export with homographic adaptation brings a significant improvement in repeatability, the second one a smaller one. For illustration, this is the repeatability improvements that I get on COCO:
- MagicPoint (no homographic adaptation): 0.673.
- SuperPoint (one round of homographic adaptation): 0.818.
- SuperPoint (two rounds of homographic adaptation): 0.845.

**Comment by ellick53** [comment]
I see, thank you.

**Comment by LZHDOPE** [comment]
@rpautrat  
Could you please send me one, your checkpoints would be very helpful!
email: [bitzhihao@163.com](url)
Thank you!

**Comment by Le0000000000n** [comment]
> I see, thank you.

Hi! I have run into similar weird problems as yours. May I ask did you solve it? Any suggestions? Thanks!

**Comment by rpautrat** [comment]
This issue should have been solved in the most recent version of the network. The problem was a resize of the image during homography adaptation that involuntary spread the the activation of the heatmap on keypoints.
